Add WorkloadManifestReader

wli3 opened the pull request:
Created by @mhutch

mhutch committed:
Add WorkloadManifestReader

Created by @mhutch

mhutch committed:
Merge manifest reader API cleanup

mhutch committed:
Rename workload readers tests & track cproj renames in sln

mhutch committed:
Add IVT for workload manifest reader tests

mhutch committed:
Make workload manifest reader errors localizable

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:
```diff
@@ -0,0 +1,28 @@
+        readonly string sdkRootPath;
+        readonly string sdkVersionBand;
+
+        public SdkDirectoryWorkloadManifestProvider(string sdkRootPath, string sdkVersionBand)
```
@mhutch it is better to just input SDK version. And the SdkDirectoryWorkloadManifestProvider will convert that into a band, or the consumer need to understand SDK version scheme

wli3 replied:
And a constructor taking `string sdkRootPath, string sdkVersionBand` makes it hard to dependency inject(i think that is the point for IWorkloadManifestProvider). Since getting sdkRootPath and sdkVersionBand is business logic. So you cannot inject a fully initialized SdkDirectoryWorkloadManifestProvider before in constructor of consuming class

wli3 replied:
sdkRootPath -> dotnetRootPath. sdk root might be considered as `program files/dotnet/sdk`. We have DOTNET_ROOT env already, so that aligns with what we want

mhutch replied:
That's fair. dotnetRootPath and sdkversion it is then.

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:
```diff
@@ -0,0 +1,212 @@
+        readonly Dictionary<string, WorkloadDefinition> workloads = new Dictionary<string, WorkloadDefinition>(StringComparer.OrdinalIgnoreCase);
+        readonly Dictionary<string, WorkloadPack> packs = new Dictionary<string, WorkloadPack>(StringComparer.OrdinalIgnoreCase);
+
+        public static WorkloadResolver Create(IWorkloadManifestProvider manifestProvider)
```
nit: Factory method is not easy to discover,  any reason it is not a constructor?

mhutch replied:
not really

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/IWorkloadManifestProvider.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public interface IWorkloadManifestProvider
+    {
+        IEnumerable<Stream> GetManifests();
```
Multiple manifests are possible?

mhutch replied:
this abstracts the process of collecting a set of workload manifests to be composed together e.g. all the workload manifests for a given sdk band

it could certainly be named better

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:
```diff
@@ -0,0 +1,212 @@
+        /// Used by MSBuild resolver to scan SDK packs for AutoImport.props files to be imported.
+        /// Used by template engine to find templatees to be added to hive.
+        /// </remarks>
+        public IEnumerable<PackInfo> GetInstalledWorkloadPacksOfKind(WorkloadPackKind kind)
```
This API has disk IO. It would be hard to test. Maybe pass in a Func for File.Exists, and if the Func is null then use the real thing?(Poor man's DI)

Normal CLI has an in memory implementation of File and Directory. But we should not add more dependencies

wli3 replied:
And packs are not accessible from the outside. Most of the functionality is not possible to test now

mhutch replied:
I added IVT so the tests could still access everything.

Maybe the tests could poke a delegate into the WorkloadResolver via an internal property, which would override the use of `File.Exists` if present.

wli3 replied:
Also make sure pack Path is constructed as

`Path.Combine(dotnetRootPath, "template-packs", pack.Id.ToLower() + "." + pack.Version.ToLower() + ".nupkg"));`

1. I verified with nuget team, `pack.Id.ToLower() + "." + pack.Version.ToLower() + ".nupkg"` is reliable to construct nupkg name from package id and package version.
2. "templates" and "template-packs" are different folders. For template packs it should be in "template-packs". "templates" is for existing non workload templates

wli3 replied:
And normalized package id should be all lower case. Although part of the system is not consistent about it

Commit:
Adding tests

Commit:
Convert TemplateLocator to use the new manifest reader

But tests has not been adjusted

wli3 commented:
@mhutch I pushed some template locator work. Pull before you make new changes

Commit:
Convert TemplateLocator to use the manifest reader

mhutch committed:
Implement workload pack enumeration and aliases

wli3 commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:
```diff
@@ -0,0 +1,28 @@
+        readonly string sdkRootPath;
+        readonly string sdkVersionBand;
+
+        public SdkDirectoryWorkloadManifestProvider(string sdkRootPath, string sdkVersionBand)
```
string sdkRootPath, string sdkVersionBand should be parameter for GetManifests. You cannot dependency inject with sdkRootPath and sdkVersionBand required in constructor

wli3 commented on src/Tests/Microsoft.NET.Sdk.WorkloadManifestReader.Tests/Manifests/Sample.json:
```diff
@@ -0,0 +1,134 @@
+            "description": "Create, build and run Android apps",
+            "kind": "dev",
+            "packs": [
+                "Xamarin.Android.Templates",
```
All nuget package name should be all lower case. Since nuget package id is normalized to all lower case. All valid manifest should not have non-lower case package id to avoid down stream mismatch.

mhutch committed:
Add test for aliased workload packs

mhutch committed:
Add Newtonsoft.Json impl of workload reader

mhutch committed:
Fix workload manifest reader namespacing casing

Commit:
Merge remote-tracking branch 'origin/release/5.0.1xx-rc2' into feature/manifest-reader

Commit:
Correct merge conflict

wli3 commented:
I pushed new changes and merge release/5.0.1xx-rc2 to resolve merge conflict. Pull before work

wli3 commented:
@mhutch plan for the next step. I'll try to add enough tests to a point I think it is acceptable to merge. And I'll directly edit the places like I commented above to save time from back and forth.

@dsplaisted you should also be able to start the convert to json manifest reader.

Once we've done that, we should merge ASAP to get a build and insert to VS just to see if RPS is still happy. Especially for msbuildresolver.

Commit:
All packs should be lower case

Since nuget package id if normalized to all lower case

Commit:
Adding tests

Commit:
Format code stype change

Commit:
Add more tests

Commit:
Add header doc

wli3 commented:
Just pushed some formatting change and tests

wli3 commented:
The loc in manifest reader is also a headache. We need to somehow merge the loc of sdk resolver and the manifest reader so that we don't have 2 set of satellite assembly

Commit:
Duplicate Strings.resx to reduce dll loading

This might be complicated. But I cannot find a better way. Problem:
workloadManifestReader has loc files. That means there will be satellite
assemblies for them. However, we want to reduce unnecessary dll loading
in msbuild resolver. It also makes template locator's packaging hard.
Since nuget cannot pack satellite assemblies of packagereferences.

Since we need to compile the source instead of package reference anyway.
I just left out String.resx. And I made a copy of Strings.resx in
template locator (appending to existing msbuild resolver's resx later).
I force these resxs to have the same namespace. So, it will pass on the
source level. The downside is whenever there is a loc change, you need
to edit 3 places -- msbuild resolver, template locator and manifest
reader. Although, the projects won't compile if we missed one. We don't
need to worry about the duplicated translation. The loc team has a
global cache. They only need to translate once.

wli3 commented:
Just did some crazy workaround:

Duplicate Strings.resx to reduce dll loading

This might be complicated. But I cannot find a better way. Problem: workloadManifestReader has loc files. That means there will be satellite assemblies for them. However, we want to reduce unnecessary dll loading in msbuild resolver. It also makes template locator's packaging hard. Since nuget cannot pack satellite assemblies of packagereferences.

Since we need to compile the source instead of package reference anyway. I just left out String.resx. And I made a copy of Strings.resx in template locator (appending to existing msbuild resolver's resx later). I force these resxs to have the same namespace. So, it will pass on the source level. The downside is whenever there is a loc change, you need to edit 3 places -- msbuild resolver, template locator and manifest reader. Although, the projects won't compile if we missed one. We don't need to worry about the duplicated translation. The loc team has a global cache. They only need to translate once.

dsplaisted commented:
Can't we just include the same .resx file in multiple projects?  I think we do that for some of the test projects where we want to check that the right strings are generated as error messages.

wli3 commented:
@dsplaisted msbuild resolver has a resx too. We cannot merge them

wli3 commented:
Talked to Mikayla C:\Program Files\dotnet\packs and C:\Program Files\dotnet\shared have cases. we cannot use all lower case. It will be bug farm. Let's try to wrap the type.

Commit:
Revert "All packs should be lower case"

This reverts commit 00404b21aa2f02a773dbd20a7e7e7f12c2eeede2.

Commit:
Revert all pack name to lower case

Commit:
Remove template locator copy of loc files

wli3 commented:
For template engine. I think it is fine to check in. Most manifest reader still don't have test coverage. But I added enough test in that code path. However, I realize resolver side of the work is still in a private branch. It is not trivial for to cover it.

wli3 commented:
@dsplaisted @mhutch let's review this and merge it asap. I pretty much directly edited my feedback

mhutch committed:
Clean up workload resolver id handling

dsplaisted commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs:
```diff
@@ -0,0 +1,306 @@
+            {
+                foreach (var workload in manifest.Workloads)
+                {
+                    _workloads.Add(workload.Key, workload.Value);
```
Will this throw if duplicate workloads or packs are defined in different workload manifests?  Should we have better error handling?

mhutch replied:
Good catch. This should still be an error IMO - right now, the approach is to fail on validation errors, in absence of a well-defined recovery strategy, but this could absolutely be a better error.

dsplaisted commented on src/Resolvers/WorkloadManifestValidator/Program.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    class Program
+    {
+        static void Main(string[] args)
```
This looks like just a stub project.  Do we need it?

mhutch replied:
Nope, I didn't get round to implementing this. It should be punted to a later PR.

dsplaisted reviewed (approved):
I didn't look much at the parsing code or the tests, but looks good.

wli3 commented:
I need to get a build soon in order to work on dotnet/installer work (a package for VS template engine to consume). I'll create a separate PR to address the code reivew

mhutch commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadManifestReader.SystemTextJson.cs:
```diff
@@ -0,0 +1,113 @@
+                var readCount = stream.Read(buffer.Memory.Span);
+                span = buffer.Memory.Slice(0, readCount).Span;
+
+                if (span.StartsWith(utf8Bom))
+                {
+                    span = span.Slice(utf8Bom.Length, span.Length - utf8Bom.Length);
+                }
```
Yes, System.Text.Json does not consider the BOM to be a valid part of UTF8 input.

> Since the Utf8JsonReader considers the input to be JSON text, a UTF-8 byte order mark (BOM) is considered invalid JSON. The caller needs to filter that out before passing the data to the reader.

https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to#read-with-a-stream-or-pipereader

mhutch commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadDefinitionkId.cs:
```diff
@@ -0,0 +1,36 @@
+    /// Wraps a workload definition id string to help ensure consistency of behaviour/semantics.
+    /// Comparisons are case insensitive but ToString() will return the original string for display purposes.
+    /// </summary>
+    internal struct WorkloadDefinitionkId : IComparable<WorkloadDefinitionkId>, IEquatable<WorkloadDefinitionkId>
```
Yes :/

mhutch commented on src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadManifestReader.NewtonsoftJson.cs:
```diff
@@ -0,0 +1,80 @@
+
+            JsonTextReader reader;
+
+            public long TokenStartIndex => reader.LineNumber; //FIXME: rationalize line/col and offset
```
Still need to fix error positions in the Newtonsoft.Json reader.
